// @pega/cosmos-react-core/lib/components/Card/Card.js
import { jsx as _jsx } from "react/jsx-runtime";
import { forwardRef } from 'react';
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../theme';
import Flex from '../Flex';
import { StyledPopover } from '../Popover';
import { StyledCardMedia } from './CardMedia';
const defaultProps = {
    interactive: false,
    role: 'region'
};
export const StyledCard = styled.article(props => {
    const { 'border-radius': borderRadius, 'border-color': borderColor, background } = props.theme.components.card;
    const interactive = props.interactive
        ? css `
        &:hover {
          border: 0;
          box-shadow: 0 0 0 0.0625rem ${borderColor};
        }
        &:focus {
          border: 0;
          box-shadow: 0 0 0 0.125rem ${borderColor};
        }
      `
        : undefined;
    return css `
    ${StyledPopover} &, &:not(& &) {
      background-color: ${background};
      border-radius: ${borderRadius};

      &:focus {
        outline: none;
      }

      ${interactive}

      ${StyledCardMedia} {
        &:first-child img {
          border-radius: ${borderRadius} ${borderRadius} 0 0;
        }
      }
    }
  `;
});
StyledCard.defaultProps = defaultThemeProp;
const Card = forwardRef((props, ref) => {
    const { children, as, onClick, interactive, role, ...restProps } = props;
    const interactiveProps = {
        ...(onClick && {
            onClick,
            tabIndex: 0,
            role: 'button'
        }),
        ...(interactive && {
            tabIndex: 0
        }),
        ...(role && { role }),
        interactive
    };
    return (_jsx(Flex, { container: { direction: 'column' }, as: StyledCard, forwardedAs: as, ref: ref, ...interactiveProps, ...restProps, children: children }));
});
Card.defaultProps = defaultProps;
export default Card;
