// @pega/cosmos-react-core/lib/components/AppShell/AppShellSearch.styles.js
import { lighten, mix } from 'polished';
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../theme';
import { tryCatch } from '../../utils';
import BareButton from '../Button/BareButton';
import { StyledLink } from '../Link';
import { StyledMenuItem } from '../Menu';
import { StyledMenu, StyledMenuListContainer } from '../Menu/Menu.styles';
import { StyledSearchInput } from '../SearchInput';
import { StyledCancelButton, StyledFiltersPopover, StyledResultsPopover, StyledSearchButton } from '../SearchInput/SearchInput.styles';
import { StyledText } from '../Text';
import { StyledBackdrop } from '../Backdrop';
import { StyledEmptyState } from '../EmptyState';
import { navWidth } from './AppShell.styles';
export const StyledSearch = styled.div(({ theme, collapsed, isMediumOrAbove }) => {
    const backgroundColor = tryCatch(() => lighten(0.1, theme.components['app-shell'].nav['background-color']));
    const maxCollapsedSize = theme.components.input.height;
    const activeColor = tryCatch(() => mix(0.85, theme.base.palette.dark, theme.base.palette.interactive));
    const hoverColor = tryCatch(() => mix(0.95, theme.base.palette.dark, theme.base.palette.interactive));
    return css `
      flex-shrink: 0;
      justify-content: center;
      --background-color: ${backgroundColor};

      ${StyledMenuListContainer} {
        background: transparent;
      }

      > form {
        padding: ${theme.base.spacing} 0;
      }

      header {
        ${StyledText} {
          color: ${theme.base.palette.light};
        }
      }

      ${StyledBackdrop} {
        background: transparent;
      }

      ${StyledEmptyState} {
        span,
        svg {
          color: ${theme.base.palette.light};
        }
      }

      ${StyledMenu} {
        &[data-active-scope='true'] [data-current='true'] {
          background-color: ${activeColor};
        }

        ${StyledLink} {
          color: ${theme.base.palette.light};
        }
      }

      ${StyledMenuItem} {
        color: ${theme.base.palette.light};

        li {
          color: ${theme.base.palette.light};
        }

        &:focus-within {
          background-color: ${activeColor};
        }

        &:hover:not([aria-disabled='true']) {
          background-color: ${hoverColor};
        }
      }

      ${StyledFiltersPopover}, ${StyledResultsPopover} {
        background-color: ${backgroundColor};
        z-index: calc(${theme.base['z-index'].backdrop} + 1);
      }

      ${StyledSearchInput} {
        border: none;
        background: transparent;

        ${StyledSearchButton} {
          transition: borderRadius calc(${theme.base.animation.speed} * 2);
          border: none;
        }

        ${StyledCancelButton} {
          color: rgba(255, 255, 255, 0.9);
        }

        input {
          border-color: ${backgroundColor};

          &::placeholder {
            color: rgba(255, 255, 255, 0.6);
          }

          &:focus:not([disabled]) {
            border-color: ${backgroundColor};
          }
        }

        input,
        ${StyledSearchButton} {
          background: ${backgroundColor};
          color: rgba(255, 255, 255, 0.7);
          border: transparent;

          &:focus-within {
            border: transparent;
          }
        }

        &:focus-within {
          ${StyledSearchButton} {
            color: rgba(255, 255, 255, 0.9);
          }
        }
      }

      form > ${StyledSearchInput} {
        margin: 0 calc((${navWidth} - ${theme.base['hit-area']['mouse-min']}) / 2);
        ${collapsed &&
        isMediumOrAbove &&
        css `
          overflow: hidden;
        `}
        background: ${backgroundColor};

        ${collapsed &&
        isMediumOrAbove &&
        css `
          ${StyledSearchButton} {
            border-radius: calc(
              ${theme.base['border-radius']} * ${theme.components['search-input']['border-radius']}
            );
            height: ${maxCollapsedSize};
            width: ${maxCollapsedSize};
            padding: 0;

            > :not(:first-child) {
              display: none;
            }
          }

          input {
            padding: 0;
            max-width: 0;
            overflow: hidden;
          }

          ${StyledCancelButton} {
            display: none;
          }
        `}
      }
    `;
});
StyledSearch.defaultProps = defaultThemeProp;
export const StyledAppShellSearchDecoration = styled.div(({ theme }) => {
    return css `
    width: ${theme.components.input.height};
    height: ${theme.components.input.height};
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    background: var(--background-color);
    color: rgba(255, 255, 255, 0.7);
    margin: 0 auto;
  `;
});
StyledAppShellSearchDecoration.defaultProps = defaultThemeProp;
export const StyledAppShellSearchButton = styled(BareButton) `
  width: 100%;
  &:focus {
    ${StyledAppShellSearchDecoration} {
      outline: none;
      box-shadow: inset ${props => props.theme.base.shadow['focus-inverted']};
    }
  }
`;
StyledAppShellSearchButton.defaultProps = defaultThemeProp;
export const StyledAppshellSearchContainer = styled.div(({ theme: { base } }) => {
    return css `
    padding: ${base.spacing};
  `;
});
StyledAppshellSearchContainer.defaultProps = defaultThemeProp;
