// @pega/cosmos-react-core/lib/components/AppShell/AppShellList.styles.js
import { lighten } from 'polished';
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../theme';
import { tryCatch } from '../../utils';
import { StyledAlert } from '../Badges/Alert';
import { StyledEmptyState } from '../EmptyState';
import { StyledSummaryItem } from '../SummaryItem';
import { StyledText } from '../Text';
/* AppShellList styles */
export const StyledAppShellListWrapper = styled.div(({ theme }) => {
    return css `
    max-height: 100%;
    white-space: nowrap;

    ${StyledSummaryItem} {
      position: relative;
      &::after {
        content: '';
        position: absolute;
        top: calc(100% - 0.03125rem);
        inset-inline-end: 1rem;
        inset-inline-start: 1rem;
        height: 0.0625rem;
        background-color: ${theme.base.colors.gray['extra-dark']};
      }
    }
  `;
});
StyledAppShellListWrapper.defaultProps = defaultThemeProp;
export const StyledAppShellList = styled.ul ``;
StyledAppShellList.defaultProps = defaultThemeProp;
export const StyledAppShellSummaryItem = styled.div(({ theme }) => {
    return css `
    padding: ${theme.base.spacing} calc(2 * ${theme.base.spacing});
    color: inherit;
    background-color: transparent;
    border: none;
    text-decoration: none;
    cursor: pointer;
    text-align: start;
    white-space: normal;

    &:hover,
    &:focus {
      color: ${theme.base.colors.white};
      background-color: rgba(255, 255, 255, 0.05);
    }

    &:focus {
      outline: none;
      box-shadow: inset ${theme.base.shadow['focus-inverted']};
    }
  `;
});
StyledAppShellSummaryItem.defaultProps = defaultThemeProp;
export const StyledAppShellMetaList = styled.ul `
  li {
    color: inherit;
  }
`;
export const StyledAppShellPinButton = styled.button(({ theme }) => {
    return css `
    color: inherit;
    border: none;
    background-color: transparent;

    &:hover,
    &:focus {
      color: ${theme.base.colors.white};
      background-color: rgba(255, 255, 255, 0.05);
    }

    &:focus {
      outline: none;
      box-shadow: inset ${theme.base.shadow['focus-inverted']};
    }
  `;
});
StyledAppShellPinButton.defaultProps = defaultThemeProp;
export const StyledNotificationVisual = styled.span(({ theme, unread, visualIncluded }) => {
    return css `
      position: relative;
      ${StyledAlert} {
        display: ${unread ? 'auto' : 'none'};
        position: absolute;
        inset-inline-start: -0.65625rem;
        top: ${visualIncluded ? 'calc(50% - 0.15625rem)' : '0.5rem'};
        background-color: ${theme.base.palette.light};
        height: 0.3125rem;
        width: 0.3125rem;
      } ;
    `;
});
StyledNotificationVisual.defaultProps = defaultThemeProp;
export const StyledNotificationPrimary = styled.span(({ theme, unread }) => {
    return css `
    font-weight: ${unread
        ? theme.base['font-weight'].bold
        : theme.base['font-weight']['semi-bold']};
    color: ${unread ? theme.base.palette.light : 'inherit'};
  `;
});
StyledNotificationPrimary.defaultProps = defaultThemeProp;
export const StyledNotificationSecondary = styled.span(({ theme, unread }) => {
    return css `
      ${StyledText} {
        opacity: ${unread
        ? theme.base.transparency['transparent-1']
        : theme.base.transparency['transparent-3']};
      }
    `;
});
StyledNotificationSecondary.defaultProps = defaultThemeProp;
export const EmptyStateColorContrast = styled.div `
  ${StyledEmptyState} {
    span,
    svg {
      color: rgba(255, 255, 255, 0.7);
    }
  }
`;
export const StyledAppShellListSection = styled.div(({ theme }) => {
    const borderColor = tryCatch(() => lighten(0.1, theme.components['app-shell'].nav['background-color']));
    return css `
    border-top: 0.0625rem solid ${borderColor};
    min-height: max(20%, 8rem);
    overflow-y: auto;
  `;
});
StyledAppShellListSection.defaultProps = defaultThemeProp;
export const StyledAppShellListHeading = styled.h3(({ theme }) => {
    return css `
    position: sticky;
    top: 0;
    padding: ${theme.base.spacing} calc(2 * ${theme.base.spacing});
    background-color: hsl(207 14% 13% / 1);
    z-index: 1;
  `;
});
StyledAppShellListHeading.defaultProps = defaultThemeProp;
