// @pega/cosmos-react-work/lib/components/Timeline/Timeline.styles.js
import styled, { css } from 'styled-components';
import { Button, calculateFontSize, Card, CardContent, DateTimeDisplay, defaultThemeProp, Flex, Grid, StyledBackdrop, StyledIcon, useDirection } from '@pega/cosmos-react-core';
export const StyledCard = styled(Card)(({ theme }) => {
    return css `
    border-top: 0.0625rem solid ${theme.base.colors.slate.light};
    padding: ${theme.base.spacing} 0;
    &:not(& &) {
      border-radius: 0;
      background-color: transparent;
    }
  `;
});
StyledCard.defaultProps = defaultThemeProp;
export const StyledCardContent = styled(CardContent)(({ theme }) => {
    return css `
    &:not(${StyledCard} ${StyledCard} > &) {
      padding: ${theme.base.spacing} 0;
      padding-inline-start: ${theme.base['hit-area']['mouse-min']};
      @media (pointer: coarse) {
        padding-inline-start: ${theme.base['hit-area']['finger-min']};
      }
    }
  `;
});
StyledCardContent.defaultProps = defaultThemeProp;
export const StyledToggleButton = styled(Button)(({ theme, collapsed }) => {
    return css `
    transition: transform calc(2 * ${theme.base.animation.speed})
      ${theme.base.animation.timing.ease};
    transform: ${collapsed ? 'rotate(-90deg)' : 'rotate(0)'};
    color: ${theme.components.button.color};
  `;
});
StyledToggleButton.defaultProps = defaultThemeProp;
export const StyledCardHeader = styled(Grid) `
  padding: 0;
  min-height: 2rem;
  ${StyledToggleButton} {
    align-self: start;
  }
`;
StyledCardHeader.defaultProps = defaultThemeProp;
export const StyledTimelineIcon = styled(Flex)(({ theme }) => {
    return css `
    border-radius: 50%;
    color: ${theme.base.palette.light};
    background: ${theme.base.palette.info};
    & > ${StyledIcon} {
      height: 1em;
      width: 1em;
    }
  `;
});
StyledTimelineIcon.defaultProps = defaultThemeProp;
export const StyledBulletWrapper = styled.td(({ theme, hasPrev, hasNext, isPrevFuture, isNextFuture, hasIcon }) => {
    const { rtl } = useDirection();
    return css `
      position: relative;
      text-align: center;
      padding-top: calc(2.25 * ${theme.base.spacing} + 0.0625rem);
      min-width: 2rem;

      ${hasPrev &&
        css `
        &::before {
          content: '';
          display: block;
          position: absolute;
          top: 0;
          inset-inline-start: 50%;
          transform: translateX(${rtl ? '50%' : '-50%'});
          width: 0.0625rem;
          height: calc(2.25 * ${theme.base.spacing} + 0.75rem);
          border-inline-start: 0.0625rem ${isPrevFuture ? 'dashed' : 'solid'}
            ${theme.base.palette.info};
        }
      `}

      ${hasNext &&
        css `
        &::after {
          content: '';
          display: block;
          position: absolute;
          top: calc(2.25 * ${theme.base.spacing} + 0.75rem);
          inset-inline-start: 50%;
          transform: translateX(${rtl ? '50%' : '-50%'});
          width: 0.0625rem;
          height: calc(100% - 2.25 * ${theme.base.spacing} - 0.75rem);
          border-inline-start: 0.0625rem ${isNextFuture ? 'dashed' : 'solid'}
            ${theme.base.palette.info};
        }
      `}

      & > ${StyledTimelineIcon} {
        position: relative;
        margin: 0;
        z-index: 1;
        width: ${hasIcon ? '1.5rem' : '0.5rem'};
        height: ${hasIcon ? '1.5rem' : '0.5rem'};
      }

      @media (min-width: ${theme.base.breakpoints.sm}) {
        padding-top: calc(${theme.base.spacing} + 0.0625rem);
        &::before {
          height: calc(${theme.base.spacing} + 0.75rem);
        }
        &::after {
          top: calc(${theme.base.spacing} + 0.75rem);
          height: calc(100% - ${theme.base.spacing} - 0.75rem);
        }
      }
    `;
});
StyledBulletWrapper.defaultProps = defaultThemeProp;
export const StyledTimelineTable = styled.table(({ theme }) => {
    return css `
    tr {
      td {
        border-inline-start: ${theme.base.spacing} solid transparent;
        border-inline-end: ${theme.base.spacing} solid transparent;
        vertical-align: top;
      }
      td:first-child {
        border-inline-start: none;
      }

      td:last-child {
        border-inline-end: none;
      }
    }
    border-collapse: collapse;
    width: 100%;
    @media (min-width: ${theme.base.breakpoints.sm}) {
      column-gap: ${theme.base.spacing};
      tr {
        td {
          border-inline-start: calc(0.5 * ${theme.base.spacing}) solid transparent;
          border-inline-end: calc(0.5 * ${theme.base.spacing}) solid transparent;
        }
      }
      ${StyledCard} {
        padding: calc(0.5 * ${theme.base.spacing}) 0;
      }
    }
  `;
});
StyledTimelineTable.defaultProps = defaultThemeProp;
export const StyledDate = styled(DateTimeDisplay)(({ theme: { base: { 'font-size': fontSize, 'font-scale': fontScale, spacing }, components: { text } } }) => {
    const fontSizes = calculateFontSize(fontSize, fontScale);
    return css `
      display: block;
      padding: calc(3 * ${spacing}) 0 calc(1.25 * ${spacing});
      font-size: ${fontSizes[text.h2['font-size']]};
      font-weight: ${text.h2['font-weight']};
    `;
});
StyledDate.defaultProps = defaultThemeProp;
export const StyledTimeWrapper = styled.td(({ theme }) => {
    return css `
    @media (min-width: ${theme.base.breakpoints.sm}) {
      text-align: end;
      line-height: ${theme.base['hit-area']['mouse-min']};
      padding-top: calc(0.5 * ${theme.base.spacing} + 0.0625rem);
    }
  `;
});
StyledTimeWrapper.defaultProps = defaultThemeProp;
export const StyledTime = styled(DateTimeDisplay)(({ theme }) => {
    return css `
    opacity: ${theme.base.transparency['transparent-2']};
    white-space: nowrap;
  `;
});
StyledTime.defaultProps = defaultThemeProp;
export const StyledLoadingCell = styled.td `
  position: relative;
  height: 2.8rem;

  ${StyledBackdrop} {
    background: transparent;
  }
`;
