// @pega/cosmos-react-core/lib/components/DateTime/Picker/DatePicker.styles.js
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../../theme';
import { StyledButton } from '../../Button';
import { StyledFormField } from '../../FormField/FormField';
import { StyledSelect } from '../../Select/Select';
export const StyledMonthPicker = styled.fieldset(props => {
    const { theme: { base: { spacing, 'border-radius': baseBorderRadius, palette: { 'foreground-color': color } }, components: { 'form-control': formControl } } } = props;
    return css `
    margin: calc(2 * ${spacing}) calc(0.5 * ${spacing});
    font-size: 0.875rem;
    border: none;

    ${StyledSelect} {
      border-radius: calc(${baseBorderRadius} * ${formControl['border-radius']});
    }

    ${StyledButton} {
      border-radius: calc(${baseBorderRadius} * ${formControl['border-radius']});
      border: 0;
      &:enabled {
        color: ${color};
      }
    }
  `;
});
StyledMonthPicker.defaultProps = defaultThemeProp;
export const StyledDatePicker = styled.div(props => {
    const { theme: { base: { spacing }, components: { 'form-control': { 'foreground-color': foreground } } } } = props;
    return css `
    width: max-content;

    ${StyledButton} {
      border: transparent;
      color: ${foreground};
    }
    padding: 0 ${spacing} ${spacing} ${spacing};

    /* Targets time input. */
    & ~ ${StyledFormField} {
      padding: 0 calc(2 * ${spacing}) calc(2 * ${spacing});
    }
  `;
});
StyledDatePicker.defaultProps = defaultThemeProp;
