// @pega/cosmos-react-core/lib/components/Badges/Keyboard.js
import { jsx as _jsx } from "react/jsx-runtime";
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../theme';
import useOS from '../../hooks/useOS';
const osxKeyMapper = {
    Command: '⌘',
    Option: '⌥',
    Alt: '⌥',
    Control: '⌃',
    Return: '↵',
    Delete: 'Del'
};
const windowsKeyMapper = {
    Control: 'Ctrl',
    Alt: 'Alt',
    Windows: '⊞',
    Backspace: 'Backspace',
    Option: 'Alt'
};
const generalKeyMapper = {
    'Page up': 'Pg up',
    'Page down': 'Pg dn',
    Ins: 'Insert'
};
export const StyledKeyboard = styled.kbd(({ theme: { base, components: { text: { primary }, badges: { keyboard } } } }) => {
    return css `
      background-color: ${keyboard['background-color']};
      color: ${base.palette['foreground-color']};
      display: inline-block;
      position: relative;
      font-family: ${base['font-family']};
      font-weight: ${primary['font-weight']};
      overflow: hidden;
      white-space: nowrap;
      padding: 0.125rem;
      border: 0.0125rem solid ${keyboard['border-color']};
      border-radius: calc(${base['border-radius']} / 4);
      text-transform: capitalize;
      text-align: center;
      min-width: 1.25rem;
      height: 1.25rem;
      text-overflow: ellipsis;
      line-height: 1rem;
      inset-block-start: calc(1.25rem / 4 - 0.125rem);
    `;
});
StyledKeyboard.defaultProps = defaultThemeProp;
const Keyboard = ({ keyName, ...restProps }) => {
    const { windows: isWindows } = useOS();
    return (_jsx(StyledKeyboard, { ...restProps, "aria-label": keyName, children: (isWindows && windowsKeyMapper[keyName]) ||
            osxKeyMapper[keyName] ||
            generalKeyMapper[keyName] ||
            keyName }));
};
export default Keyboard;
