// @pega/custom-dx-components/src/constants.js
// cSpell:ignore words timeofday caseview casepreview objectview
export const TASKS_CONFIG_JSON_FILENAME = 'tasks.config.json';
export const PACKAGE_JSON_FILENAME = 'package.json';
export const BUILD_CONFIG_JSON_FILENAME = 'build.config.json';
export const WEBPACK_CONFIG_JS_FILENAME = 'webpack.config.js';
export const WEBPACK_PATHS_JS_FILENAME = 'paths.js';
export const COMPONENTS_DIRECTORY_PATH = 'components-directory-path';
export const COMPONENTS_PATH = 'component';
export const TOKEN_PATH = 'node_modules/@pega/custom-dx-components/.access_token';
export const OOTB_COMPONENTS = 'node_modules/@pega/custom-dx-components/.ootb_components';

/* REST end points */
export const PUBLISH_COMPONENT_SERVICE_REST_ENDPOINT = 'PRRestService/dev/v1/publish_component';
export const LIST_COMPONENT_SERVICE_REST_ENDPOINT = 'PRRestService/dev/v1/list_component';
export const OOTB_COMPONENT_SERVICE_REST_ENDPOINT = 'PRRestService/dev/v1/ootb_component';
export const DELETE_COMPONENT_SERVICE_REST_ENDPOINT = 'PRRestService/dev/v1/delete_component';
export const INF_TOKEN_URL_V1 = 'PRRestService/oauth2/v1/token';
export const INF_AUTHORIZE_URL_V1 = 'PRRestService/oauth2/v1/authorize';
export const LP_TOKEN_URL = 'uas/oauth/token';
export const LP_AUTHORIZE_URL = 'uas/oauth/authorize';
export const LP_LIST_COMPONENT_SERVICE_REST_ENDPOINT = 'pegalaunchpad/c11n/v103/app/appstudiox/componentslist?field0=type&value0={type}';
export const LP_PUBLISH_COMPONENT_SERVICE_REST_ENDPOINT = 'pegalaunchpad/c11n/isolations/{isolationId}/v102/assetsUpdate';
export const LP_DELETE_COMPONENT_SERVICE_REST_ENDPOINT = 'pegalaunchpad/c11n/isolations/{isolationId}/v102/assetsUpdate';

export const FIELD_PATH = '/field';
export const TEMPLATE_PATH = '/template';
export const WIDGET_PATH = '/widget';

export const COMPONENT_PREFIX_FIELD_PATH = 'field/';
export const COMPONENT_PREFIX_TEMPLATE_PATH = 'template/';
export const COMPONENT_PREFIX_WIDGET_PATH = 'widget/';

export const BUILDER_INDEX_FILE_PATH = '../bundle/builder/index.js';
export const BUILDER_DIR_FILE_PATH = '../bundle/builder';

export const CATEGORY_CONSTELLATION = 'constellation';

export const AUTHENTICATE_DEFAULTS = {
  serverType: 'infinity',
  server: 'https://localhost:1080/prweb',
  clientId: '10095186356008396159',
  grantType: 'authCode',
  authService: 'pega',
  redirectUri: 'https://localhost:4010/'
};

export const AUTHENTICATE_SCHEMA = {
  serverType: [
    {
      name: 'Infinity',
      value: 'infinity'
    },
    {
      name: 'Launchpad',
      value: 'launchpad'
    }
  ],
  grantType: [
    {
      name: 'Authorization code',
      value: 'authCode'
    },
    {
      name: 'Password credentials',
      value: 'passwordCreds'
    },
    {
      name: 'Client credentials',
      value: 'clientCreds'
    }
  ]
};

export const COMPONENT_SCHEMA = {
  framework: [
    {
      name: 'Constellation',
      value: 'Constellation'
    },
    {
      name: 'SDK Angular',
      value: 'SDK-Angular'
    },
    {
      name: 'SDK React',
      value: 'SDK-React'
    },
    {
      name: 'SDK Web Components',
      value: 'SDK-WebComponents'
    }
  ],
  type: [
    {
      name: 'Field',
      value: 'Field'
    },
    {
      name: 'Layout template',
      value: 'Template'
    },
    {
      name: 'Widget',
      value: 'Widget'
    }
  ],
  subtype: {
    field: [
      {
        name: 'Text',
        value: 'Text'
      },
      {
        name: 'Text Input',
        value: 'Text-Input'
      },
      {
        name: 'Paragraph',
        value: 'Text-Paragraph'
      },
      {
        name: 'Email',
        value: 'Text-Email'
      },
      {
        name: 'Phone',
        value: 'Text-Phone'
      },
      {
        name: 'URL',
        value: 'Text-URL'
      },
      {
        name: 'Integer',
        value: 'Integer'
      },
      {
        name: 'Decimal',
        value: 'Decimal'
      },
      {
        name: 'Currency',
        value: 'Decimal-Currency'
      },
      {
        name: 'Percentage',
        value: 'Decimal-Percentage'
      },
      {
        name: 'Boolean',
        value: 'Boolean'
      },
      {
        name: 'Date',
        value: 'Date'
      },
      {
        name: 'DateTime',
        value: 'DateTime'
      },
      {
        name: 'TimeOfDay',
        value: 'TimeOfDay'
      },
      {
        name: 'Picklist',
        value: 'Picklist'
      }
    ],
    template: [
      {
        name: 'DETAILS',
        value: 'DETAILS'
      },
      {
        name: 'DETAILS region',
        value: 'DETAILS-region'
      },
      {
        name: 'FORM',
        value: 'FORM'
      },
      {
        name: 'FORM region',
        value: 'FORM-region'
      },
      {
        name: 'PAGE',
        value: 'PAGE'
      }
    ],
    widget: [
      {
        name: 'CASE',
        value: 'CASE'
      },
      {
        name: 'PAGE',
        value: 'PAGE'
      },
      {
        name: 'PAGE & CASE',
        value: ['PAGE', 'CASE']
      }
    ]
  }
};
