// @pega/cosmos-react-work/lib/components/CaseView/UtilitiesSummary.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled, { css } from 'styled-components';
import { Card, Count, Flex, Icon, useI18n } from '@pega/cosmos-react-core';
export const StyledUtilitiesSummary = styled.div(({ theme }) => {
    return css `
    width: 2.75rem;
    padding: ${theme.base.spacing} 0;
    cursor: pointer;
  `;
});
const UtilitiesSummary = ({ items = [], onClick }) => {
    const t = useI18n();
    return items.length ? (_jsx(Card, { interactive: true, tabIndex: 0, role: 'button', container: { direction: 'column', alignItems: 'center', gap: 3 }, "aria-label": `${t('utilities_summary')}. ${t('expand_utilities_panel')}`, as: StyledUtilitiesSummary, onClick: onClick, onKeyDown: (e) => {
            if (e.key === ' ' || e.key === 'Enter')
                onClick?.();
        }, children: items.map(item => {
            return (_jsxs(Flex, { container: {
                    direction: 'column',
                    alignItems: 'center',
                    gap: 0.5
                }, item: { grow: 1 }, children: [_jsx(Icon, { name: item.iconName }), item.count !== undefined && _jsx(Count, { variant: 'default', children: item.count })] }, item.iconName));
        }) })) : null;
};
export default UtilitiesSummary;
