// @pega/cosmos-react-core/lib/components/DateTime/Input/QuarterInput.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, useReducer } from 'react';
import styled from 'styled-components';
import FormField from '../../FormField';
import Select, { Option } from '../../Select';
import Flex from '../../Flex';
import { range } from '../../../utils';
import { useConfiguration, useConsolidatedRef, useFocusWithin, useI18n } from '../../../hooks';
import { StyledSelect } from '../../Select/Select';
import { defaultThemeProp } from '../../../theme';
import { StyledFormControl } from '../../FormControl';
import { getQuarter, parseToDate } from './utils';
import StyledDateTime from './DateTime.styles';
export const StyledQuarterInput = styled(StyledDateTime) `
  padding: 0;
  border: 0;

  ${StyledSelect} {
    min-width: 0;

    &:focus {
      z-index: 1;
    }

    &:first-child {
      border-start-end-radius: 0;
      border-end-end-radius: 0;
    }

    &:last-child {
      max-width: max-content;
      margin-inline-start: -0.0625rem;
      border-start-start-radius: 0;
      border-end-start-radius: 0;
    }
  }
`;
StyledQuarterInput.defaultProps = defaultThemeProp;
const convertToCallbackParameter = (date, parts) => {
    if ([parts.quarter, parts.year].every(Boolean)) {
        const quarterStartMonth = [0, 3, 6, 9][Number(parts.quarter) - 1];
        const resultDate = new Date(date);
        resultDate.setUTCFullYear(Number(parts.year), quarterStartMonth, 1);
        return {
            valueAsISOString: resultDate.toISOString(),
            valueAsTimestamp: resultDate.getTime()
        };
    }
    if ([parts.year, parts.quarter].every(p => !p)) {
        return {
            valueAsISOString: '',
            valueAsTimestamp: undefined
        };
    }
    return {
        valueAsISOString: '',
        valueAsTimestamp: NaN,
        state: 'incomplete'
    };
};
const reducer = (state, action) => {
    if (action.field) {
        return { ...state, [action.field]: action.value };
    }
    return state;
};
const QuarterInput = forwardRef((props, ref) => {
    const { defaultValue, value = defaultValue, min, max, id, label, labelHidden, info, status, required, readOnly, disabled, onChange, onFocus, onBlur, autoFocus, additionalInfo, ...restProps } = props;
    const { locale } = useConfiguration();
    const date = value ? parseToDate(value) : undefined;
    const currentYear = date ? date.getUTCFullYear() : new Date().getFullYear();
    const minYear = min ? parseToDate(min).getUTCFullYear() : currentYear - 10;
    const maxYear = max ? parseToDate(max).getUTCFullYear() : currentYear + 10;
    const years = range(minYear, maxYear);
    const [state, dispatch] = useReducer(reducer, {
        year: date?.getUTCFullYear().toString(),
        quarter: date && getQuarter(date).toString()
    });
    const onFocusChange = (focused) => {
        const callbackParam = convertToCallbackParameter(date ? new Date(date) : new Date(), state);
        if (onFocus && focused)
            onFocus(callbackParam);
        else if (onBlur && !focused)
            onBlur(callbackParam);
    };
    const onSelectChange = (field) => (ev) => {
        const action = { field, value: ev.target.value };
        dispatch(action);
        onChange?.(convertToCallbackParameter(date ? new Date(date) : new Date(), reducer(state, action)));
    };
    const containerRef = useConsolidatedRef(ref);
    useFocusWithin([containerRef], onFocusChange);
    const t = useI18n();
    const quarters = [1, 2, 3, 4];
    const quarterOptions = quarters.map(quarterNumber => (_jsx(Option, { value: quarterNumber.toString(), children: t(`date_quarter_q${quarterNumber}`).trim() }, t(`date_quarter_q${quarterNumber}`).trim())));
    const yearOptions = years.map(year => (_jsx(Option, { value: year.toString(), children: year.toString() }, year.toString())));
    const displayNames = new Intl.DisplayNames(locale, { style: 'long', type: 'dateTimeField' });
    const Quarter = (_jsxs(Select, { "aria-label": displayNames.of('quarter'), readOnly: readOnly, required: required, value: state.quarter?.toString() ?? '', onChange: onSelectChange('quarter'), status: status, disabled: disabled, autoFocus: autoFocus, children: [!required && _jsx(Option, { children: " " }, 'null'), quarterOptions] }, 'quarter'));
    const Year = (_jsxs(Select, { "aria-label": displayNames.of('year'), readOnly: readOnly, required: required, value: state.year?.toString() ?? '', onChange: onSelectChange('year'), status: status, disabled: disabled, children: [!required && _jsx(Option, { children: " " }, 'null'), yearOptions] }, 'year'));
    const Comp = (_jsxs(Flex, { ...restProps, as: StyledQuarterInput, forwardedAs: StyledFormControl, container: { alignItems: 'center' }, ref: containerRef, status: status, disabled: disabled, readOnly: readOnly, children: [Quarter, Year] }));
    return label ? (_jsx(FormField, { as: 'fieldset', labelAs: 'legend', ...{ label, labelHidden, id, info, status, required, disabled, additionalInfo }, children: Comp })) : (Comp);
});
export default QuarterInput;
